libshpool/src/daemon/prompt.rs

```diff
@@ -15,94 +15,334 @@
+        // this rather than `echo $PROMPT_SENTINEL` because different
+        // shells have subtly different echo behavior which makes it
+        // hard to make the scanner work right.
+        // TODO(julien): this will probably not work on mac
+        let sentinel_cmd =
+            format!("\n{}=yes /proc/{}/exe daemon\n", PROMPT_SENTINEL_FLAG_VAR, std::process::id());
+        script.push_str(sentinel_cmd.as_str());
+
+        let mut pty_master = pty_master.is_parent().context("expected parent")?;
+        pty_master.write_all(script.as_bytes()).context("running prefix script")?;
+
+        Ok(())
+    }
+
+    /// Determine the real shell being executed by the given command. You might
+    /// think that if a user executes `/bin/bash` they are starting up a
+    /// bash shell. While this is usually true, it isn't always the case
+    /// thanks to `exec`. Though the official way to change your shell is
+    /// via chsh, sysadmins often mess with this for various reasons, and
+    /// this can often mean that users of more exotic shells prefer to rely
+    /// on good ol `exec` rather than trying to deal with whatever
+    /// strange variant of a system shell configuration setup they have been
+    /// given.
+    ///
+    /// In the face of arbitrary code execution being used for configuring which
+    /// shell the user wants to use, all we can do is execute the code, then try
+    /// to pick up the peices later. We do this by creating a throw-away
+    /// subshell and then echoing a `ps` command bracketed in some sentinals
+    /// in order to determine who is really running.
+    ///
+    /// Note that we don't need to worry about this elsewhere because stuff
+    /// like the SHELL variable still points to the users default shell even
+    /// if they are using exec to switch to a different shell.
+    #[instrument(skip_all)]
+    fn shell_of_command(&mut self, shell: &str) -> anyhow::Result<KnownShell> {
+        debug!("DEBUGTRACE 1");
+        if let Some(s) = self.true_shells.get(shell) {
+            debug!("DEBUGTRACE 2");
+            return s.clone().ok_or(anyhow!("no known shell for '{}'", shell));
+        }
+        debug!("DEBUGTRACE 3");
+
+        let shell_basename = Path::new(&shell)
+            .file_name()
+            .ok_or(anyhow!("error building login shell indicator"))?
+            .to_str()
+            .ok_or(anyhow!("error parsing shell name as utf8"))?;
+        debug!("DEBUGTRACE 4");
+        let mut shell_proc = process::Command::new(shell)
+            .stdin(process::Stdio::piped())
+            .stdout(process::Stdio::piped())
+            .stderr(process::Stdio::piped())
+            .arg0(format!("-{}", shell_basename))
+            .spawn()?;
+        debug!("DEBUGTRACE 5");
+
+        let mut sniff_cmd =
+            format!(r#"echo "{}$(ps -p {} -o comm=){}""#, SNIFF_START, shell_proc.id(), SNIFF_END);
+        sniff_cmd.push('\n');
```

Comment:
just to be sure i understand, this doesn't help prevent issues where the user's `bash` binary has been replaced with fish?
this would be fine with aliases though correct? (tested `ps -p ... -o comm=` locally seems to be true)

Reply:
outdated due to rewriting to get tests passing